core/src/main/java/org/apache/iceberg/encryption/DefaultEncryptionManagerFactory.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.PropertyUtil;
+
+public class DefaultEncryptionManagerFactory implements EncryptionManagerFactory {
+  public static final String ICEBERG_ENCRYPTION_TABLE_KEY = "iceberg.encryption.table.key.id";
+
+  private static class KmsClientSupplier {
+    private final boolean ignoreTableProperties;
+
+    KmsClientSupplier(boolean ignoreTableProperties) {
+      this.ignoreTableProperties = ignoreTableProperties;
+    }
+
+    KeyManagementClient create(Map<String, String> encryptionProperties) {
+      // load kms impl from catalog properties, if not present fall back to table properties (if
+      // allowed).
+      String kmsImpl = encryptionProperties.get(CatalogProperties.ENCRYPTION_KMS_CLIENT_IMPL);
+
+      Preconditions.checkArgument(
+          null != kmsImpl,
+          "KMS Client implementation class is not set (via "
+              + CatalogProperties.ENCRYPTION_KMS_CLIENT_IMPL
+              + " catalog property nor "
+              + TableProperties.ENCRYPTION_KMS_CLIENT_IMPL
+              + " table property). "
+              + "IgnoreTableProperties is: "
+              + ignoreTableProperties);
+
+      return KmsUtil.loadKmsClient(kmsImpl, encryptionProperties);
+    }
+  }
+
+  private KmsClientSupplier kmsClientSupplier;
+  private KeyManagementClient client;
+  private Map<String, String> catalogPropertyMap;
+  private boolean ignoreTableProperties;
+
+  @Override
+  public void initialize(Map<String, String> catalogProperties) {
+    this.catalogPropertyMap = catalogProperties;
+    ignoreTableProperties =
+        PropertyUtil.propertyAsBoolean(
+            catalogProperties,
+            CatalogProperties.ENCRYPTION_IGNORE_TABLE_PROPS,
+            CatalogProperties.ENCRYPTION_IGNORE_TABLE_PROPS_DEFAULT);
+    kmsClientSupplier = new KmsClientSupplier(ignoreTableProperties);
+  }
+
+  @Override
+  public EncryptionManager create(TableMetadata tableMetadata) {
+    if (null == tableMetadata) {
+      return PlaintextEncryptionManager.INSTANCE;
+    }
+
+    Map<String, String> tableProperties = tableMetadata.properties();
+    String fileFormat =
+        PropertyUtil.propertyAsString(tableProperties, DEFAULT_FILE_FORMAT, "parquet");
+    if (!fileFormat.equals("parquet")) {
+      throw new UnsupportedOperationException(
+          "Iceberg encryption currently supports only parquet format for data files");
+    }
+
+    final Map<String, String> encryptionProperties = Maps.newHashMap();
+    if (!ignoreTableProperties) {
+      encryptionProperties.putAll(tableProperties);
+    }
+
+    // Important: put catalog properties after table properties. Former overrides the latter.
+    encryptionProperties.putAll(catalogPropertyMap);
+
+    String tableKeyId = System.getProperty(ICEBERG_ENCRYPTION_TABLE_KEY);
```

Comment:
This seems odd to me.
Do we rely on `System` properties?

Reply:
This is basically a hack that allows to address a challenge of using different keys for different tables - in those cases where the table properties are not trusted. I'd say the purpose is somewhere between a "temporary solution" and a "discussion trigger" :). In the long run, we might have table properties safely kept in catalogs (eg the REST catalog), but we need a solution in a shorter term.